packages/block-editor/src/components/line-height-control/index.js

```diff
@@ -8,17 +8,19 @@ import { ZERO } from '@wordpress/keycodes';
 /**
  * Internal dependencies
  */
+import useEditorFeature from '../use-editor-feature';
 import {
 	BASE_DEFAULT_VALUE,
 	RESET_VALUE,
 	STEP,
-	useIsLineHeightControlsDisabled,
 	isLineHeightDefined,
 } from './utils';
 
 export default function LineHeightControl( { value: lineHeight, onChange } ) {
 	// Don't render the controls if disabled by editor settings
-	const isDisabled = useIsLineHeightControlsDisabled();
+	const isDisabled = useEditorFeature(
+		'__experimentalDisableCustomLineHeight'
+	);
```

Comment:
I like the API a lot, and the fact that we can implement any heuristics to get the value of that config in the hook without changing how the API is used.

I'm not sure yet about the naming. `useEditorFeature`, `useEditorSetting`, `useEditorConfig` ? How will it scale to nested configs like typography, colors? Why is `__experimentalDisableCustomLineHeight` a top level config and not nested under "typography" (same for drop cap).

cc @mtias @aduth @mcsf and others.

Reply:
> I like the API a lot, and the fact that we can implement any heuristics to get the value of that config in the hook without changing how the API is used.

🤘

Reply:
> I'm not sure yet about the naming. `useEditorFeature`, `useEditorSetting`, `useEditorConfig` ?

If the thing we're referring to here is the same thing we assign via `<Editor settings={ ... } />`, or `block_editor_settings` filter, I'd very much like to see "setting" or "settings" as part of the name.

>How will it scale to nested configs like typography, colors? Why is `__experimentalDisableCustomLineHeight` a top level config and not nested under "typography" (same for drop cap).

Do we have nested settings? Can we not? 😛

Reply:
If we fly with `useSetting` then we need those nested groups to make it all easier. We discussed groups like `typography`, `colors` so far. I'm sure we need more. The only question is whether we name it feature, setting, support or whatever :)

Reply:
for context, this hook is supposed to get its values from the "config" property in the proposed format in the end. https://github.com/WordPress/gutenberg/issues/21583#issuecomment-613911005

This config is probably going to end up in the `block-editor` getSettings selector somehow, I haven't give it much though yet about how, whether it's just `1 - 1` or if there's some intermediate mapping that happens.

I don't think the block-editor settings and the "config" of a block-editor are the exact same things, block-editor can have more settings and themes can have more "config" though.

Reply:
It feels like all we need is proper namespacing to make it easy to inject all settings from the `theme.json` config. Wherever we have duplication, we can find a nice way to migrate it behind the scenes before sending it to the front.

To give an example, the following would still work on PHP side:

```php
$settings = array(
    'disableCustomColors'    => get_theme_support( 'disable-custom-colors' ),
    'disableCustomFontSizes' => get_theme_support( 'disable-custom-font-sizes' ),		
);
```

You could also override it with `theme.json`:

```json
{ 
    "config": {
        "colors": {
            "disableCustomColors": false
        },
        "typography": {
            "disableCustomFontSizes": false
        }
    }
}
```

The challenge would be to pick what takes precedence, `theme.json` or `get_theme_support` here, or the top-level `disableCustomFontSizes` vs `typography. disableCustomFontSizes'. Before sending it to the client, we would remove duplicates from the top-level using some conflict resolution strategy.

On the client, we need to keep old settings for backward compatibility but I would solve it by doing internal mapping in the `getSettings` selector:

```js
const mappings = {
    disableCustomColors: 'colors.disableCustomColors',
    disableCustomFontSizes: 'typography.disableCustomFontSizes',
};
```

I hope I didn't miss anything.